### DISCUSSION TITLE
Nested Monte Carlo Analysis

### DISCUSSION THREAD

**Discussion by jessroche** [snl-dakota/dakota#64] [Q&A] [answered]
Hi, I'm fairly new to Dakota and have only used it for simple sensitivity analyses so far. I am interested in running a nested Monte Carlo sensitivity analysis where a set of variables are sampled in the outer loop and for each of those samples, a different set of variables are sampled in the inner loop. For example with 5 samples in each loop, the resulting values for each sample would look like the table below. Samples 0-5 are sampled in the outer loop and samples 4-6 are sampled in the inner loop, using the same values from the outer loop for all 5 samples in the inner loop. I looked into the multilevel sampling method, but it doesn't seem to fit my needs to specify different variables for each level. What would be the best way to do this in Dakota?

Thanks,
Jessica

sample_0 | sample_1 | sample_2 | sample_3 | sample_4 | sample_5 | sample_6
-- | -- | -- | -- | -- | -- | --
10.382 | 0.993 | 5.954 | 3.53E+07 | 0.657 | 0.559 | 0.696
10.382 | 0.993 | 5.954 | 3.53E+07 | 0.858 | 0.127 | 0.855
10.382 | 0.993 | 5.954 | 3.53E+07 | 0.932 | 0.358 | 0.179
10.382 | 0.993 | 5.954 | 3.53E+07 | 0.843 | 0.012 | 0.086
10.382 | 0.993 | 5.954 | 3.53E+07 | 0.672 | 0.580 | 0.948
10.198 | 0.996 | 4.978 | 3.52E+07 | 0.602 | 0.326 | 0.824
10.198 | 0.996 | 4.978 | 3.52E+07 | 0.721 | 0.943 | 0.323
10.198 | 0.996 | 4.978 | 3.52E+07 | 0.446 | 0.360 | 0.943
10.198 | 0.996 | 4.978 | 3.52E+07 | 0.380 | 0.969 | 0.539
10.198 | 0.996 | 4.978 | 3.52E+07 | 0.689 | 0.348 | 0.220
10.969 | 1.014 | 7.705 | 1.87E+07 | 0.793 | 0.752 | 0.633
10.969 | 1.014 | 7.705 | 1.87E+07 | 0.656 | 0.120 | 0.841
10.969 | 1.014 | 7.705 | 1.87E+07 | 0.311 | 0.582 | 0.561
10.969 | 1.014 | 7.705 | 1.87E+07 | 0.284 | 0.098 | 0.259
10.969 | 1.014 | 7.705 | 1.87E+07 | 0.907 | 0.764 | 0.333
9.074 | 1.015 | 6.595 | 4.72E+06 | 0.474 | 0.530 | 0.115
9.074 | 1.015 | 6.595 | 4.72E+06 | 0.343 | 0.866 | 0.373
9.074 | 1.015 | 6.595 | 4.72E+06 | 0.243 | 0.893 | 0.533
9.074 | 1.015 | 6.595 | 4.72E+06 | 0.680 | 0.532 | 0.521
9.074 | 1.015 | 6.595 | 4.72E+06 | 0.618 | 0.544 | 0.190
9.140 | 1.019 | 6.393 | 1.29E+07 | 0.548 | 0.004 | 0.918
9.140 | 1.019 | 6.393 | 1.29E+07 | 0.026 | 0.381 | 0.020
9.140 | 1.019 | 6.393 | 1.29E+07 | 0.352 | 0.373 | 0.924
9.140 | 1.019 | 6.393 | 1.29E+07 | 0.164 | 0.228 | 0.241
9.140 | 1.019 | 6.393 | 1.29E+07 | 0.307 | 0.304 | 0.649

**Comment by jadamstephens** [comment, marked as answer]
Jessica,

You need a "nested" study. They can be a little tricky to set up. I'm providing an example in this comment. The outer sampling method performs ten random samples of the variable x1. For each, the inner sampling method draws ten samples of x2 while holding x1 constant. The inner method computes the mean and standard deviation of the response f. These are mapped into the response that belongs to the outer method: mean_f = 1.0 * mean + 0.0 * stddev.

Something that might trip you up is that the tabular file produced by this study will contain evaluation information only for the outer method. That is: samples of x1 and the mean of f. If you want ALL the samples and corresponding values of your inner study response (i.e. the one from your simulation), you'll have to get them from Dakota's [HDF5 output](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/output/hdf/organizationofevaluations.html), scrape them from the console output, or generate them from the parameters and results files.


You should also look at the keyword documentation for [nested model](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/model-nested.html). Pay particular attention to the [sub_method_pointer](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/model-nested-sub_method_pointer.html) documentation.


```
environment
  method_pointer 'outer_method'
  tabular_data

method
  id_method 'outer_method'
  sampling
    seed 1337
    samples 10
  model_pointer 'nested'

model
  id_model 'nested'
  nested
  variables_pointer 'outer_v'
  responses_pointer 'outer_r'
  sub_method_pointer 'inner_method'
    primary_response_mapping 1 0
    primary_variable_mapping 'x1'

variables
  id_variables 'outer_v'
  uniform_uncertain 1
    descriptors 'x1'
    lower_bounds -1
    upper_bounds  1

responses
  id_responses 'outer_r'
  response_functions 1
    descriptors 'mean_f'
    no_gradients
    no_hessians

method
  id_method 'inner_method'
  sampling
    seed 8675309
    samples 10
  model_pointer 'sim'

model
  id_model 'sim'
  simulation
  variables_pointer 'inner_v'
  responses_pointer 'inner_r'

variables
  id_variables 'inner_v'
  active uncertain
  continuous_state 1
    descriptors 'x1'
  uniform_uncertain 1
    descriptors 'x2'
    lower_bounds -1
    upper_bounds  1

responses
  id_responses 'inner_r'
  response_functions  1
    descriptors 'f'
    no_gradients
    no_hessians

interface
  analysis_drivers 'text_book'
  direct
```

**Reply by jessroche** [reply]
Thank you for the detailed response. I was able to use this example to set up an input file that perfectly suits my needs.
